processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java

```diff
@@ -107,35 +172,134 @@ public void close() throws IOException
         headerOut.getCount()
     );
     Preconditions.checkState(
-        numBytesWritten < Integer.MAX_VALUE, "Wrote[%s] bytes, which is too many.", numBytesWritten
+        numBytesWritten < fileSizeLimit, "Wrote[%s] bytes, which is too many.", numBytesWritten
     );
 
-    OutputStream metaOut = ioPeon.makeOutputStream(makeFilename("meta"));
-
-    try {
+    try (OutputStream metaOut = ioPeon.makeOutputStream(makeFilename("meta"))) {
       metaOut.write(0x1);
       metaOut.write(objectsSorted ? 0x1 : 0x0);
       metaOut.write(Ints.toByteArray((int) numBytesWritten + 4));
-      metaOut.write(Ints.toByteArray(numWritten));
+      metaOut.write(Ints.toByteArray((int) numWritten));
+    }
+  }
+
+  private void closeMultiFiles() throws IOException
+  {
+    headerOut.close();
+    headerOutLong.close();
+    valuesOut.close();
+
+    // revisit this check.
+    Preconditions.checkState(
+        headerOutLong.getCount() == (numWritten * 8),
+        "numWritten[%s] number of rows should have [%s] bytes written to headerOutLong, had[%s]",
+        numWritten,
+        numWritten * 8,
+        headerOutLong.getCount()
+    );
+    Preconditions.checkState(
+        headerOutLong.getCount() < Integer.MAX_VALUE,
+        "Wrote[%s] bytes in header file, which is too many.",
+        headerOutLong.getCount()
+    );
+  }
+
+  /**
+   * Tries to get best value split(number of elements in each value file) which can be expressed as power of 2.
+   *
+   * @return Returns the size of value file splits as power of 2.
+   *
+   * @throws IOException
+   */
+  private int bagSizePower() throws IOException
+  {
+    int avgObjectSize = (int) (valuesOut.getCount() / numWritten);
+
+    File f = ioPeon.getFile(makeFilename("headerLong"));
+    Preconditions.checkNotNull(f, "header file missing.");
+
+    try (RandomAccessFile headerFile = new RandomAccessFile(f, "r")) {
+      for (int i = 31; i >= 0; --i) {
+        if ((1L << i) * avgObjectSize <= fileSizeLimit) {
+          if (actuallyFits(i, headerFile)) {
+            return i;
+          }
+        }
+      }
     }
-    finally {
-      metaOut.close();
+    throw new ISE(
```

Comment:
say a druid user got this error during batch/realtime ingestion , it will continue to happen on retries ... what would be the resolution?